### PR TITLE
🚨🚨 Source Instagram: upgrade to v18

### LINKED ISSUE #33899: Source Instagram: migrate to latest API (v18)
### Connector Name

source-instagram

### Connector Version

2.0.0

### What step the error happened?

None

### Relevant information

Reason: https://github.com/airbytehq/oncall/issues/3875

Migrate to latest API (v18).
Will affect existing metrics, and therefore, schemas, so it will be a breaking change improvement.

# TODO:

remove unused (OLD) metrics (see https://github.com/airbytehq/airbyte/pull/33889), change to relevant for all streams 

### Relevant log output

_No response_

### Contribute

- [ ] Yes, I want to contribute

### PR DESCRIPTION
**Pull request by artem1205**
## What

Resolve https://github.com/airbytehq/airbyte/issues/33899

## How

upgrade to v18

## Recommended reading order

1. `airbyte-integrations/connectors/source-instagram/setup.py`
2. `streams.py`

## 🚨 User Impact 🚨

Reauthentication for all users required, scope added: `business_management` 
see https://github.com/airbytehq/airbyte-platform-internal/pull/10582 

- Stream Media Insights

  | Old metric                 | New metric         |
  |----------------------------|--------------------|
  | carousel_album_engagement  | total_interactions |
  | carousel_album_impressions | impressions        |
  | carousel_album_reach       | reach              |
  | carousel_album_saved       | saved              | 
  | carousel_album_video_views | video_views        |
  | engagement                 | total_interactions |

  New metrics for Reels: `ig_reels_avg_watch_time`, `ig_reels_video_view_total_time`

- Stream UserLifetimeInsights

    Metric `audience_locale` become unavailable.
    Metrics `audience_city`, `audience_country` and `audience_gender_age` changed to one metric `follower_demographics` with respective breakdowns `city`, `country`, `age,gender`.
    primary key changed to ["business_account_id", "breakdown"]
    date property deleted.
    Metric type set to `total_value`, see [docs](https://developers.facebook.com/docs/instagram-api/reference/ig-user/insights#metric-type) for more info.

- Stream StoryInsights
  deprecated metrics: `exits`, `taps_forward`, `taps_back`

## Pre-merge Actions

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-instagram/integration_tests/expected_records.jsonl [hunk 1]
```diff
@@ -2,3 +2,6 @@
...
+{"stream":"user_lifetime_insights","data":{"page_id":"144706962067225","breakdown":"city","business_account_id":"17841408147298757","metric":"follower_demographics"},"emitted_at":1704378481116}
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-instagram/integration_tests/expected_records.jsonl:5]: Why were these added? It seems to me like the API should have been returning these even before the change so I'm probably missing some information to understand

File: airbyte-integrations/connectors/source-instagram/source_instagram/streams.py [hunk 2]
```diff
@@ -143,10 +143,18 @@ def read_records(
 class UserLifetimeInsights(DatetimeTransformerMixin, InstagramStream):
     """Docs: https://developers.facebook.com/docs/instagram-api/reference/ig-user/insights"""
 
-    primary_key = ["business_account_id", "metric", "date"]
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-instagram/source_instagram/streams.py:146]: I don't have any sample response from the previous version so I can't confirm what `date` really was. However, if the value was being updated every sync, it would mean that there was a new entry created in the destination every time. This will not be the case anymore. Is this an issue? Should we document this behavior change?

**Review by maxi297 — COMMENTED** [review]
This seems on the right track but I lack information to approve. Can you answer those questions?

File: airbyte-integrations/connectors/source-instagram/integration_tests/expected_records.jsonl (hunk 1 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-instagram/integration_tests/expected_records.jsonl:5]: I removed this in the previous PR because no records were returned for v11. However, no error was raised.

File: airbyte-integrations/connectors/source-instagram/source_instagram/streams.py (hunk 2 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-instagram/source_instagram/streams.py:146]: You could find previous version [here](https://github.com/airbytehq/airbyte/blob/cc7f01925397c7d773f8cae33eb19651bf697c85/airbyte-integrations/connectors/source-instagram/integration_tests/expected_records.jsonl)
This is an issue, because data will be returned for ALL period, see https://github.com/airbytehq/airbyte/pull/33930/files#diff-ade20df2f7ea130a82ed265a6a96eb70c0d9ccb98e92bb61afdb5b05c146525fR27

File: docs/integrations/sources/instagram-migrations.md [hunk 3]
```diff
@@ -1,5 +1,59 @@
...
+  Old metric will be replaced with the new ones, refer to the [IG Media Insights](https://developers.facebook.com/docs/instagram-api/reference/ig-media/insights#metrics) for more info.
+
+  | Old metric                 | New metric         |
+  |----------------------------|--------------------|
+  | carousel_album_engagement  | total_interactions |
+  | carousel_album_impressions | impressions        |
+  | carousel_album_reach       | reach              |
+  | carousel_album_saved       | saved              | 
+  | carousel_album_video_views | video_views        |
+  | engagement                 | total_interactions |
```
[the remark this answers is not in the document]
Comment (artem1205) [reply docs/integrations/sources/instagram-migrations.md:18]: Agreed, note added

File: docs/integrations/sources/instagram-migrations.md [hunk 4]
```diff
@@ -1,5 +1,59 @@
...
+
+  New metrics for Reels: `ig_reels_avg_watch_time`, `ig_reels_video_view_total_time`
```
[the remark this answers is not in the document]
Comment (artem1205) [reply docs/integrations/sources/instagram-migrations.md:20]: not a breaking change, added just to gather all available data.

File: docs/integrations/sources/instagram-migrations.md [hunk 5]
```diff
@@ -1,5 +1,59 @@
...
+
+    - Metric `audience_locale` will become unavailable.
+    - Metrics `audience_city`, `audience_country`, and `audience_gender_age` will be consolidated into a single metric named `follower_demographics`, featuring respective breakdowns for `city`, `country`, and `age,gender`.
+    - Primary key will be changed to `["business_account_id", "breakdown"]`.
+    - The metric type will be set to 'total_value'. For more information, refer to the [docs](https://developers.facebook.com/docs/instagram-api/reference/ig-user/insights#metric-type) for more info.
+
+
+- `Story Insights`
+
+  Metrics: `exits`, `taps_back`, `taps_forward` will become unavailable.
```
[the remark this answers is not in the document]
Comment (artem1205) [reply docs/integrations/sources/instagram-migrations.md:32]: Agreed, I've seen this, but decided to leave "old" metrics as is in stream (seems like they are not compatible with breakdown, it is only my suggestions, can not test it, we have no data). Instagram `insights` are very similar to FB marketing now, that's why I'll create a follow up issue to add custom report builder in our configuration.

[the remark this answers is not in the document]
Comment (artem1205) [reply docs/integrations/sources/instagram-migrations.md:38]: @maxi297 , follow up issue: https://github.com/airbytehq/airbyte/issues/33984

File: airbyte-integrations/connectors/source-instagram/source_instagram/streams.py (hunk 2 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-instagram/source_instagram/streams.py:146]: So basically, we are moving from `metric_type=time_series` to `metric_type= total_value`, right? If so, is this what the customer wants? This seems like a different stream entirely

Comment (artem1205) [reply airbyte-integrations/connectors/source-instagram/source_instagram/streams.py:146]: Not really, we'll receive the [most `similar` metrics](https://github.com/airbytehq/airbyte/pull/33930/files#diff-ade20df2f7ea130a82ed265a6a96eb70c0d9ccb98e92bb61afdb5b05c146525fR31) as it was described in [FB docs (alternative metric)](https://developers.facebook.com/docs/instagram-api/reference/ig-user/insights#metrics-and-periods)

Comment (maxi297) [reply airbyte-integrations/connectors/source-instagram/source_instagram/streams.py:146]: Yes but they are not over time right?

Comment (artem1205) [reply airbyte-integrations/connectors/source-instagram/source_instagram/streams.py:146]: No, they are "lifetime", just 1 record per metric without any date.

Comment (maxi297) [reply airbyte-integrations/connectors/source-instagram/source_instagram/streams.py:146]: Now yes, but before, "date" was part of the primary key right? So we could have multiple entries in the database for the same metric (depending on the date)

Comment (maxi297) [reply airbyte-integrations/connectors/source-instagram/source_instagram/streams.py:146]: Sync discussion summary: The metric does not support `time_series`. So this is a breaking change and we don't have an easy way around it. We're fine with the change and will try to intercept user feedback if this is not good enough. In the meanwhile, @artem1205 will document the change better in the migration guide.

**Review by maxi297 — APPROVED** [review]
LGTM! :shipit:

File: airbyte-integrations/connectors/source-instagram/metadata.yaml [hunk 6]
```diff
@@ -21,6 +21,13 @@ data:
   releaseStage: generally_available
   releases:
     breakingChanges:
+      3.0.0:
+        message: "Connector will be upgraded to API v18. Customers who use streams `Media Insights`, `Story Insights` or `User Lifetime Insights` must take action with their connections."
```
Comment (katmarkham) [inline airbyte-integrations/connectors/source-instagram/metadata.yaml:25]: The existing Instagram API (v11) has been deprecated. Customers who use streams `Media Insights`, `Story Insights` or `User Lifetime Insights` must take action with their connections.  Please follow the <migration guide> to update to the latest Instagram API (v18).

File: docs/integrations/sources/instagram-migrations.md [hunk 7]
```diff
@@ -1,5 +1,70 @@
 # Instagram Migration Guide
 
+## Upgrading to 3.0.0
+
+Connector will be upgraded to API v18. Affected Streams and their corresponding changes are listed below:
```
Comment (katmarkham) [inline docs/integrations/sources/instagram-migrations.md:5]: The Instagram connector has been upgrade to API v18 (following the deprecation of v11).  Connector will be upgraded to API v18. Affected Streams and their corresponding changes are listed below:

File: docs/integrations/sources/instagram-migrations.md [hunk 8]
```diff
@@ -1,5 +1,70 @@
...
+  | carousel_album_engagement  | total_interactions |
+  | carousel_album_impressions | impressions        |
+  | carousel_album_reach       | reach              |
+  | carousel_album_saved       | saved              | 
+  | carousel_album_video_views | video_views        |
+  | engagement                 | total_interactions |
+
+:::note
+
+You may see different results: `engagement` includes likes, comments and saves count while `total_interactions` includes likes, comments, saved and shares count.
```
Comment (katmarkham) [inline docs/integrations/sources/instagram-migrations.md:22]: You may see different results: `engagement` count includes likes, comments, and saves while `total_interactions` count includes likes, comments, and saves, as well as shares.

**Review by katmarkham — CHANGES_REQUESTED** [review]
please add link to migration guide and make other changes requested

File: airbyte-integrations/connectors/source-instagram/metadata.yaml (hunk 6 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-instagram/metadata.yaml]: updated

**Review by katmarkham — APPROVED** [review]
